packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts

```diff
@@ -0,0 +1,436 @@
+        performAggregatedSlippageCheck ? 0n : amountOut,
+        path,
+        payerIsUser,
+      ]
+      planner.addCommand(CommandType.V3_SWAP_EXACT_IN, exactInputSingleParams)
+    } else {
+      const exactOutputSingleParams: ABIParametersType<CommandType.V3_SWAP_EXACT_OUT> = [
+        recipient,
+        amountOut,
+        amountIn,
+        path,
+        payerIsUser,
+      ]
+      planner.addCommand(CommandType.V3_SWAP_EXACT_OUT, exactOutputSingleParams)
+    }
+  }
+}
+
+// encode a mixed route swap, i.e. including both v2 and v3 pools
+async function addMixedSwap(
+  planner: RoutePlanner,
+  trade: SmartRouterTrade<TradeType>,
+  options: PancakeSwapOptions,
+  payerIsUser: boolean,
+  routerMustCustody: boolean,
+  performAggregatedSlippageCheck: boolean,
+): Promise<void> {
+  const isExactIn = trade.tradeType === TradeType.EXACT_INPUT
+
+  for (const route of trade.routes) {
+    const { inputAmount, outputAmount, pools } = route
+    const amountIn: bigint = SmartRouter.maximumAmountIn(trade, options.slippageTolerance).quotient
+    const amountOut: bigint = SmartRouter.minimumAmountOut(trade, options.slippageTolerance).quotient
+
+    // flag for whether the trade is single hop or not
+    const singleHop = pools.length === 1
+
+    const recipient = routerMustCustody ? ROUTER_AS_RECIPIENT : validateAndParseAddress(options.recipient!)
+
+    const mixedRouteIsAllV3 = (r: Omit<BaseRoute, 'input' | 'output'>) => {
+      return r.pools.every(SmartRouter.isV3Pool)
+    }
+    const mixedRouteIsAllV2 = (r: Omit<BaseRoute, 'input' | 'output'>) => {
+      return r.pools.every(SmartRouter.isV2Pool)
+    }
+
+    const mixedRouteIsAllStable = (r: Omit<BaseRoute, 'input' | 'output'>) => {
+      return r.pools.every(SmartRouter.isStablePool)
+    }
+
+    // similar to encodeMixedRouteSwap but more simplified where we just continue
+    // as if its a regular v2 or v3 trade
+    if (singleHop) {
+      if (mixedRouteIsAllV3(route)) {
+        addV3Swap(
+          planner,
+          {
+            ...trade,
+            routes: [route],
+            inputAmount,
+            outputAmount,
+          },
+          options,
+          routerMustCustody,
+          payerIsUser,
+          performAggregatedSlippageCheck,
+        )
+      } else if (mixedRouteIsAllV2(route)) {
+        addV2Swap(
+          planner,
+          {
+            ...trade,
+            routes: [route],
+            inputAmount,
+            outputAmount,
+          },
+          options,
+          routerMustCustody,
+          payerIsUser,
+          performAggregatedSlippageCheck,
+        )
+      } else if (mixedRouteIsAllStable(route)) {
+        addStableSwap(
+          planner,
+          {
+            ...trade,
+            routes: [route],
+            inputAmount,
+            outputAmount,
+          },
+          options,
+          routerMustCustody,
+          performAggregatedSlippageCheck,
+          payerIsUser,
+        )
```

Comment:
seems like the input arguments are the same. Let's combine them to avoid repetition

Reply:
ohh yeah thats acxtually good idea to simplify. i could aslo do same in actual smart router too as the code there is similar i think. ill do for both i can

Reply:
ended up not doing this as the arguments are seperate and not one object, so i cant make a variable of the args to use to spread, unless i changed ll the addSwap functions to take an object of params which i dont think is worth doing on all three just to save a small bit of dup code. however if u really want comment again and i will do